keras/utils/data_utils.py

```diff
@@ -477,26 +493,43 @@ def start(self, workers=1, max_queue_size=10):
             self.executor = multiprocessing.Pool(workers)
         else:
             self.executor = ThreadPool(workers)
+        self.workers = workers
         self.queue = queue.Queue(max_queue_size)
         self.stop_signal = threading.Event()
         self.run_thread = threading.Thread(target=self._run)
         self.run_thread.daemon = True
         self.run_thread.start()
 
+    def wait_queue(self):
+        """Wait for the queue to be empty."""
+        while True:
+            time.sleep(0.1)
+            if self.queue.unfinished_tasks == 0 or self.stop_signal.is_set():
+                return
+
     def _run(self):
         """Function to submit request to the executor and queue the `Future` objects."""
         sequence = list(range(len(self.sequence)))
+        self._send_sequence()  # Share the initial sequence
         while True:
             if self.shuffle:
                 random.shuffle(sequence)
             for i in sequence:
                 if self.stop_signal.is_set():
                     return
                 self.queue.put(
-                    self.executor.apply_async(get_index,
-                                              (self.sequence, i)), block=True)
+                    self.executor.apply_async(get_index, (self.uid, i)), block=True)
+
+            # Done with the current epoch, waiting for the final batches
+            self.wait_queue()
+
+            if self.stop_signal.is_set():
```

Comment:
Returning here means that we're done, so we do not call the on_epoch_end of the Sequence to save time.
So the validation Sequence cannot use on_epoch_end. Is that alright?

Reply:
In validation or test, there is no concept of epoch, so that seems OK to me.